File: Design-Patterns-Using-Ruby/Proxy/ProxyTest.rb
# Subject interface
module Subject
    def request
        raise NotImplementedError, "Subclasses must implement the 'request' method"
    end
end

# RealSubject class
class RealSubject
    include Subject

    def request
        puts 'Concrete Subject Request Method'
    end
end

# Proxy class
class Proxy
    include Subject

    def initialize
        @real_subject = RealSubject.new
    end

    def request
        @real_subject.request
    end
end

# Client code
proxy = Proxy.new
proxy.request

=begin 
Concrete Subject Request Method
=end

File: Design-Patterns-Using-JavaScript/Proxy/ProxyTest.js
// Interface
class Subject {
    request() {
        throw new Error('request method must be implemented');
    }
}

// RealSubject class
class RealSubject extends Subject {
    request() {
        console.log("Concrete Subject Request Method");
    }
}

// Proxy class
class Proxy extends Subject {
    constructor() {
        super();
        this.realSubject = new RealSubject();
    }

    request() {
        this.realSubject.request();
    }
}

// Client code
const proxy = new Proxy();
proxy.request();


/*
Concrete Subject Request Method
*/